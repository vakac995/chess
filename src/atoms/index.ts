// Core authentication atoms
export {
  authAtom,
  userAtom,
  isAuthenticatedAtom,
  authErrorAtom,
  authStatusAtom,
  loginAtom,
  logoutAtom,
} from './Auth.atoms';

// Authentication form atoms
export {
  registrationStepAtom,
  basicInfoAtoms,
  personalInfoAtoms,
  loginFormAtoms,
  registrationFormAtoms,
  isRegistrationCompleteAtom,
} from './Authentication.atoms';

// Calendar atoms
export {
  calendarAtom,
  calendarModeAtom,
  calendarConfigAtom,
  selectedDateAtom,
  selectedRangeAtom,
  calendarSelectionAtom,
  calendarErrorAtom,
  calendarStatusAtom,
  calendarLoadingAtom,
  isRangeCompleteAtom,
  hasSelectionAtom,
  selectSingleDateAtom,
  selectDateRangeAtom,
  clearCalendarSelectionAtom,
  resetCalendarAtom,
  initializeCalendarAtom,
  isDateSelectedAtom,
  isDateInRangeAtom,
} from './Calendar.atoms';

// Types
export type { User } from './Auth.atoms';
export type { LoginFormData as LoginFormAtomData } from './Authentication.atoms';
