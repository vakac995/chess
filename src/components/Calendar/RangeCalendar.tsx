import React from 'react';
import { DayPicker } from 'react-day-picker';
import { useAtom, useAtomValue } from 'jotai';
import type { DateRange } from 'react-day-picker';
import {
  selectedRangeAtom,
  selectDateRangeAtom,
  calendarConfigAtom,
  isRangeCompleteAtom,
} from '@/atoms';
import type { RangeCalendarProps } from './Calendar.types';
import {
  validateCalendarProps,
  getStandardizedDayPickerClasses,
  mergeAccessibilityProps,
  validateCalendarState,
} from './Calendar.utils';
import { WarningInfo } from './WarningInfo';

/**
 * Range selection calendar component (single calendar view)
 * Uses react-day-picker in range selection mode with a single calendar
 * Includes proper accessibility and validation features
 */
export const RangeCalendar = React.forwardRef<HTMLDivElement, RangeCalendarProps>((props, ref) => {
  // Validate props at the top
  const validation = validateCalendarProps(props);
  // All hooks and prop destructuring must be at the top
  const [, selectRange] = useAtom(selectDateRangeAtom);
  const currentSelection = useAtomValue(selectedRangeAtom);
  const config = useAtomValue(calendarConfigAtom);
  const isRangeComplete = useAtomValue(isRangeCompleteAtom);
  const warnings = validation.warnings;
  const {
    selected,
    defaultSelected,
    onSelect,
    min,
    max,
    className,
    classNames,
    styles,
    startMonth,
    endMonth,
    disabled,
    hidden,
    numberOfMonths = 1,
    showOutsideDays = true,
    fixedWeeks = true,
    showWeekNumber = false,
    modifiers,
    modifiersClassNames,
    modifiersStyles,
    dir,
    lang,
    autoFocus,
    onMonthChange,
    onNextClick,
    onPrevClick,
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledBy,
    'aria-describedby': ariaDescribedBy,
    role: _role,
    tabIndex,
    mode: _,
    ...restProps
  } = props;

  // Use controlled selection if provided, otherwise use atom state
  const currentEffectiveSelection = selected ?? currentSelection;

  // Defensive copy for DayPicker
  const dayPickerSelectedProp = currentEffectiveSelection
    ? {
        from: currentEffectiveSelection.from
          ? new Date(
              currentEffectiveSelection.from.getFullYear(),
              currentEffectiveSelection.from.getMonth(),
              currentEffectiveSelection.from.getDate()
            )
          : undefined,
        to: currentEffectiveSelection.to
          ? new Date(
              currentEffectiveSelection.to.getFullYear(),
              currentEffectiveSelection.to.getMonth(),
              currentEffectiveSelection.to.getDate()
            )
          : undefined,
      }
    : undefined;

  React.useEffect(() => {
    if (process.env.NODE_ENV === 'development' && warnings.length > 0) {
      console.warn('RangeCalendar validation warnings:', warnings);
    }
  }, [warnings]);

  const handleRangeSelect = React.useCallback(
    (proposedRange: DateRange | undefined, clickedDate: Date) => {
      let rangeToUpdate: DateRange | undefined = proposedRange;
      if (proposedRange?.from && proposedRange?.to) {
        const from = proposedRange.from < proposedRange.to ? proposedRange.from : proposedRange.to;
        const to = proposedRange.from < proposedRange.to ? proposedRange.to : proposedRange.from;
        const diffTime = Math.abs(to.getTime() - from.getTime());
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24)) + 1;
        if ((min !== undefined && diffDays < min) || (max !== undefined && diffDays > max)) {
          rangeToUpdate = { from: clickedDate, to: clickedDate };
        }
      }
      selectRange(rangeToUpdate);
      onSelect?.(rangeToUpdate);
    },
    [selectRange, onSelect, min, max]
  );

  const calendarClassNames = React.useMemo(() => {
    const standardClasses = getStandardizedDayPickerClasses();
    const rootClasses = [standardClasses.root];
    if (classNames?.root) rootClasses.push(classNames.root);
    if (className) rootClasses.push(className);
    return {
      ...standardClasses,
      ...classNames,
      root: rootClasses.join(' ').trim(),
    };
  }, [className, classNames]);

  const calendarStyles = React.useMemo(
    () => ({
      root: { margin: 0 },
      ...styles,
    }),
    [styles]
  );

  const accessibilityProps = React.useMemo(() => {
    const baseProps = {
      'aria-label': ariaLabel ?? 'Select a date range',
      'aria-labelledby': ariaLabelledBy,
      'aria-describedby':
        ariaDescribedBy ?? (!isRangeComplete ? 'range-incomplete-hint' : undefined),
    };
    return mergeAccessibilityProps(baseProps, 'range-single');
  }, [ariaLabel, ariaLabelledBy, ariaDescribedBy, isRangeComplete]);

  const selectedDaysCount = React.useMemo(() => {
    if (!currentEffectiveSelection?.from || !currentEffectiveSelection?.to) return 0;
    const diffTime = Math.abs(
      currentEffectiveSelection.to.getTime() - currentEffectiveSelection.from.getTime()
    );
    return Math.ceil(diffTime / (1000 * 60 * 60 * 24)) + 1;
  }, [currentEffectiveSelection]);

  // State validation
  const [stateValidation, setStateValidation] = React.useState<{
    isValid: boolean;
    errors: string[];
  }>({ isValid: true, errors: [] });

  React.useEffect(() => {
    const state = {
      mode: 'range-single' as const,
      selectedDate: undefined,
      selectedRange: currentEffectiveSelection,
    };
    const result = validateCalendarState(state);
    setStateValidation({ isValid: result.isValid, errors: result.errors as string[] });
    if (!result.isValid && process.env.NODE_ENV === 'development') {
      console.error('Calendar state validation error:', result.errors);
    }
  }, [currentEffectiveSelection]);

  // Handle validation errors after all hooks
  if (!validation.isValid) {
    return (
      <div
        ref={ref}
        className="calendar-validation-error border-destructive bg-destructive/10 rounded-lg border p-4"
        role="alert"
        aria-live="polite"
      >
        <h3 className="text-destructive mb-2 text-sm font-medium">Calendar Configuration Error</h3>
        <ul className="text-destructive/80 space-y-1">
          {validation.errors.map((error, index) => (
            <li key={`error-${error.slice(0, 20)}-${index}`} className="text-sm">
              • {error}
            </li>
          ))}
        </ul>
      </div>
    );
  }

  if (!stateValidation.isValid) {
    return (
      <div
        ref={ref}
        className="calendar-validation-error border-destructive bg-destructive/10 rounded-lg border p-4"
        role="alert"
        aria-live="polite"
      >
        <h3 className="text-destructive mb-2 text-sm font-medium">Calendar State Error</h3>
        <ul className="text-destructive/80 space-y-1">
          {stateValidation.errors.map((error, index) => (
            <li key={`state-error-${String(error).slice(0, 20)}-${index}`} className="text-sm">
              • {error}
            </li>
          ))}
        </ul>
      </div>
    );
  }
  return (
    <div
      ref={ref}
      className={`range-calendar-container ${className ?? ''}`.trim()}
      tabIndex={tabIndex}
      {...accessibilityProps}
    >
      {/* Show warnings if any */}
      <WarningInfo warnings={warnings} />
      <DayPicker
        mode="range"
        selected={dayPickerSelectedProp}
        onSelect={handleRangeSelect}
        defaultMonth={defaultSelected?.from}
        startMonth={startMonth}
        endMonth={endMonth}
        disabled={disabled}
        hidden={hidden}
        numberOfMonths={numberOfMonths}
        showOutsideDays={showOutsideDays}
        fixedWeeks={fixedWeeks}
        showWeekNumber={showWeekNumber}
        weekStartsOn={config.weekStartsOn}
        classNames={calendarClassNames}
        styles={calendarStyles}
        modifiers={modifiers}
        modifiersClassNames={modifiersClassNames}
        modifiersStyles={modifiersStyles}
        dir={dir}
        lang={lang}
        autoFocus={autoFocus}
        onMonthChange={onMonthChange}
        onPrevClick={onPrevClick}
        onNextClick={onNextClick}
        {...restProps}
      />
      {/* Range selection status */}
      {currentEffectiveSelection?.from && !currentEffectiveSelection?.to && (
        <div className="mt-2">
          <p className="text-muted-foreground text-xs" id="range-incomplete-hint">
            Select end date to complete range
          </p>
        </div>
      )}
      {/* Selected range info */}
      {isRangeComplete && (
        <div className="mt-2">
          <p className="text-muted-foreground text-xs">
            {selectedDaysCount} day{selectedDaysCount !== 1 ? 's' : ''} selected
          </p>
        </div>
      )}
      {/* Range constraints info */}
      {!isRangeComplete && (min !== undefined || max !== undefined) && (
        <div className="mt-2 space-y-1">
          {min !== undefined && max !== undefined && (
            <p className="text-muted-foreground text-xs" data-testid="min-max-constraint-message">
              Select between {min} and {max} days
            </p>
          )}
          {min !== undefined && max === undefined && (
            <p className="text-muted-foreground text-xs" data-testid="min-constraint-message">
              Select at least {min} day{min !== 1 ? 's' : ''}
            </p>
          )}
          {min === undefined && max !== undefined && (
            <p className="text-muted-foreground text-xs" data-testid="max-constraint-message">
              Select at most {max} day{max !== 1 ? 's' : ''}
            </p>
          )}
        </div>
      )}
    </div>
  );
});

RangeCalendar.displayName = 'RangeCalendar';
