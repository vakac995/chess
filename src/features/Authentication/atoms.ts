import { z } from 'zod';
import { atom } from 'jotai';
import { createFormAtom, combineFormAtoms, FormAtomReturn } from '@/hooks/useJotaiForm';
import { loginSchema, BasicInfoData, PersonalInfoData, RegistrationData } from './schemas';
import { LoadingStatus } from '@/types/status';

export type LoginFormData = z.infer<typeof loginSchema>;

export const loginFormAtoms = createFormAtom<LoginFormData>();

// Registration step tracking
export const registrationStepAtom = atom<1 | 2>(1);

// Registration atoms
export const basicInfoAtoms = createFormAtom<BasicInfoData>();
export const personalInfoAtoms = createFormAtom<PersonalInfoData>();

// Registration state (derived atoms)
export const isRegistrationCompleteAtom = atom<boolean>(get => {
  const basicInfoStatus = get(basicInfoAtoms.formStatusAtom);
  const personalInfoStatus = get(personalInfoAtoms.formStatusAtom);

  return (
    basicInfoStatus === LoadingStatus.FULFILLED && personalInfoStatus === LoadingStatus.FULFILLED
  );
});

// Combined registration data
export const registrationFormAtoms = combineFormAtoms<RegistrationData>(
  basicInfoAtoms as FormAtomReturn<unknown>,
  personalInfoAtoms as FormAtomReturn<unknown>
);
