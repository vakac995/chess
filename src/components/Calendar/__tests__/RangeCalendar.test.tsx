import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, act, renderHook } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { useSetAtom } from 'jotai';
import { RangeCalendar } from '../RangeCalendar';
import { initializeCalendarAtom, clearCalendarSelectionAtom } from '@/atoms';
import type { RangeCalendarProps } from '../Calendar.types';
import type { DayPickerProps, DateRange } from 'react-day-picker';

// Helper functions for the mock DayPicker component
const createMockDayPickerHelpers = (props: DayPickerProps & { mode: 'range'; selected?: DateRange }) => {
  // Helper to check if a date is disabled
  const isDateDisabled = (date: Date) => {
    if (props.disabled) {
      if (typeof props.disabled === 'function') {
        return props.disabled(date);
      } else if (Array.isArray(props.disabled)) {
        return props.disabled.some(d => {
          if (d instanceof Date) {
            return (
              d.getFullYear() === date.getFullYear() &&
              d.getMonth() === date.getMonth() &&
              d.getDate() === date.getDate()
            );
          }
          return false;
        });
      }
    }
    return false;
  };

  // Helper to handle date selection
  const handleDateClick = (date: Date) => {
    if (isDateDisabled(date)) return;

    const currentSelected = props.selected;
    let newSelection: DateRange | undefined;

    if (!currentSelected?.from) {
      // No selection, start new range
      newSelection = { from: date, to: undefined };
    } else if (!currentSelected.to) {
      // Has 'from', now set 'to'
      if (date < currentSelected.from) {
        newSelection = { from: date, to: currentSelected.from };
      } else {
        newSelection = { from: currentSelected.from, to: date };
      }
    } else {
      // Complete range exists, start new selection
      newSelection = { from: date, to: undefined };
    }

    // Call onSelect with the new selection and the clicked date
    if (props.onSelect) {
      // Mock the modifiers and event objects to match the expected signature
      const mockModifiers = {};
      const mockEvent = {
        type: 'click',
        preventDefault: () => {},
        stopPropagation: () => {},
      } as React.MouseEvent;
      props.onSelect(newSelection, date, mockModifiers, mockEvent);
    }
  };

  // Helper to handle keyboard events
  const handleKeyDown = (event: React.KeyboardEvent, date: Date) => {
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault();
      handleDateClick(date);
    }
  };

  return { isDateDisabled, handleDateClick, handleKeyDown };
};

// Helper to create a calendar day cell
const createCalendarDayCell = (
  day: number,
  dayIndex: number,
  year: number,
  month: number,
  props: DayPickerProps & { mode: 'range'; selected?: DateRange },
  helpers: ReturnType<typeof createMockDayPickerHelpers>
) => {
  const date = new Date(year, month, day);
  const isDisabled = helpers.isDateDisabled(date);
  const isSelected =
    (props.selected?.from &&
      props.selected.from.getFullYear() === date.getFullYear() &&
      props.selected.from.getMonth() === date.getMonth() &&
      props.selected.from.getDate() === date.getDate()) ||
    (props.selected?.to &&
      props.selected.to.getFullYear() === date.getFullYear() &&
      props.selected.to.getMonth() === date.getMonth() &&
      props.selected.to.getDate() === date.getDate());

  return (
    <td key={dayIndex}>
      <button
        aria-label={date.toDateString()}
        data-testid={`date-${day}`}
        disabled={isDisabled}
        data-selected={isSelected}
        onClick={() => helpers.handleDateClick(date)}
        onKeyDown={e => helpers.handleKeyDown(e, date)}
        tabIndex={day === 1 ? 0 : -1}
      >
        {day}
      </button>
    </td>
  );
};

// Mock the DayPicker component
vi.mock('react-day-picker', async importOriginal => {
  const actual = await importOriginal();
  const actualExports = typeof actual === 'object' && actual !== null ? actual : {};

  const mockDayPicker = vi.fn((props: DayPickerProps & { mode: 'range'; selected?: DateRange }) => {
    // Generate a grid of clickable dates for the given month
    const startMonth = props.startMonth || new Date(2025, 5, 1); // Default to June 2025
    const year = startMonth.getFullYear();
    const month = startMonth.getMonth();
    const daysInMonth = new Date(year, month + 1, 0).getDate();

    const helpers = createMockDayPickerHelpers(props);

    return (
      <div
        data-testid="mocked-daypicker"
        aria-label={
          props.captionLayout === 'dropdown'
            ? 'Month dropdown'
            : (props.startMonth?.toLocaleString('default', { month: 'long', year: 'numeric' }) ??
              'Calendar grid')
        }
        data-start-month={props.startMonth?.toISOString()}
        data-selected-from={props.selected?.from?.toISOString()}
        data-selected-to={props.selected?.to?.toISOString()}
        data-classnames-root={props.classNames?.root}
        role="grid"
      >
        <div>
          Mocked DayPicker. Displayed Month:{' '}
          {props.startMonth?.toLocaleString('default', { month: 'long', year: 'numeric' }) ?? 'N/A'}
          {props.selected?.from && <span>Selected From: {props.selected.from.toDateString()}</span>}
          {props.selected?.to && <span>Selected To: {props.selected.to.toDateString()}</span>}
        </div>
        <table aria-label="Calendar grid">
          <tbody>
            {Array.from({ length: Math.ceil(daysInMonth / 7) }, (_, weekIndex) => (
              <tr key={weekIndex}>
                {Array.from({ length: 7 }, (_, dayIndex) => {
                  const day = weekIndex * 7 + dayIndex + 1;
                  if (day > daysInMonth) {
                    return <td key={dayIndex}></td>;
                  }

                  return createCalendarDayCell(day, dayIndex, year, month, props, helpers);
                })}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  });

  return {
    ...actualExports,
    DayPicker: mockDayPicker,
  };
});

// Get a reference to the mocked DayPicker
import { DayPicker } from 'react-day-picker';
const mockDayPicker = vi.mocked(DayPicker);

// After mocking, importing DayPicker will give us the mock function directly.
// We can then use dayPickerMockFn for assertions.

const defaultTestProps: RangeCalendarProps = {
  mode: 'range-single',
  startMonth: new Date(2025, 5, 1), // June 2025
  onSelect: vi.fn(),
  classNames: {},
};

describe('RangeCalendar', () => {
  beforeEach(() => {
    if (mockDayPicker && typeof mockDayPicker.mockClear === 'function') {
      mockDayPicker.mockClear();
    }

    const { result: initResult } = renderHook(() => useSetAtom(initializeCalendarAtom));
    act(() => {
      initResult.current({ mode: 'range-single' });
    });
    const { result: clearResult } = renderHook(() => useSetAtom(clearCalendarSelectionAtom));
    act(() => {
      clearResult.current();
    });

    // Clear mock after state setup to avoid counting initialization renders
    if (mockDayPicker && typeof mockDayPicker.mockClear === 'function') {
      mockDayPicker.mockClear();
    }
  });

  it('renders without crashing', () => {
    render(<RangeCalendar {...defaultTestProps} />);
    expect(mockDayPicker).toHaveBeenCalled();

    // Check that DayPicker was called with correct basic props
    const firstCall = mockDayPicker.mock.calls[0];
    if (firstCall && firstCall[0]) {
      expect(firstCall[0]).toMatchObject({
        startMonth: defaultTestProps.startMonth,
        mode: 'range',
      });
    }

    const mockElement = screen.queryByTestId('mocked-daypicker');
    expect(mockElement).toBeInTheDocument();
    if (mockElement) {
      expect(mockElement).toHaveAttribute('aria-label', 'June 2025');
      expect(mockElement).toHaveTextContent(/Mocked DayPicker. Displayed Month: June 2025/);
    }
  });

  it('applies custom className to the DayPicker component via classNames.root prop', () => {
    const customClass = 'custom-range-calendar';
    const classNamesRoot = 'other-rdp-root-class';
    render(
      <RangeCalendar
        {...defaultTestProps}
        className={customClass} // This is for the RangeCalendar's own wrapper div
        classNames={{ root: classNamesRoot, month_caption: 'test-month_caption-class' }} // This is for DayPicker
      />
    );

    // Check that RangeCalendar's wrapper div gets the className prop
    const mockedDayPicker = screen.queryByTestId('mocked-daypicker');
    const rangeCalendarWrapper = mockedDayPicker?.parentElement;
    expect(rangeCalendarWrapper).toHaveClass('range-calendar-container');

    // Verify what DayPicker mock receives for classNames.root
    expect(mockDayPicker).toHaveBeenCalled();
    const lastCall = mockDayPicker.mock.calls[mockDayPicker.mock.calls.length - 1];
    expect(lastCall[0].classNames?.root).toContain(classNamesRoot);
    expect(lastCall[0].classNames?.root).toContain(customClass);
    expect(lastCall[0].classNames?.month_caption).toBe('test-month_caption-class');

    const mockElement = screen.getByTestId('mocked-daypicker');
    // The mock renders `props.classNames.root` directly for easy checking
    const rootClassAttr = mockElement.getAttribute('data-classnames-root');
    expect(rootClassAttr).toContain(classNamesRoot);
    expect(rootClassAttr).toContain(customClass);
  });

  it('displays a pre-selected range', () => {
    const initialFrom = new Date(2025, 5, 5);
    const initialTo = new Date(2025, 5, 10);
    const startMonth = new Date(2025, 5, 1);

    render(
      <RangeCalendar
        mode="range-single"
        numberOfMonths={1}
        startMonth={startMonth}
        selected={{ from: initialFrom, to: initialTo }}
        onSelect={() => {}}
      />
    );

    expect(mockDayPicker).toHaveBeenCalled();
    // Get the last call since component may re-render
    const lastCallIndex = mockDayPicker.mock.calls.length - 1;
    const callArgs = mockDayPicker.mock.calls[lastCallIndex][0] as DayPickerProps & {
      mode: 'range';
      selected?: DateRange;
    };
    expect(callArgs.selected).toEqual(
      expect.objectContaining({
        from: expect.any(Date),
        to: expect.any(Date),
      })
    );
    if (callArgs.selected?.from && callArgs.selected?.to) {
      // Compare normalized local dates (the component normalizes dates to local midnight)
      expect(callArgs.selected.from.toISOString().substring(0, 10)).toBe(
        new Date(2025, 5, 5).toISOString().substring(0, 10)
      );
      expect(callArgs.selected.to.toISOString().substring(0, 10)).toBe(
        new Date(2025, 5, 10).toISOString().substring(0, 10)
      );
    } else {
      throw new Error('Selected from/to dates were not passed to DayPicker as expected.');
    }

    const mockElementWithSelection = screen.getByTestId('mocked-daypicker');
    expect(mockElementWithSelection).toHaveTextContent(
      `Selected From: ${initialFrom.toDateString()}`
    );
    expect(mockElementWithSelection).toHaveTextContent(`Selected To: ${initialTo.toDateString()}`);
    // Verify the mock element displays the correct month
    expect(mockElementWithSelection).toHaveTextContent('Displayed Month: June 2025');
  });

  // Test for clearing selection by passing undefined to 'selected' prop
  it('clears selection when selected prop is undefined', () => {
    const initialFrom = new Date(2025, 5, 5);
    const initialTo = new Date(2025, 5, 10);
    const startMonth = new Date(2025, 5, 1);

    const { rerender } = render(
      <RangeCalendar
        mode="range-single"
        startMonth={startMonth}
        selected={{ from: initialFrom, to: initialTo }}
        onSelect={() => {}}
      />
    );

    expect(mockDayPicker).toHaveBeenCalled();
    let lastCallIndex = mockDayPicker.mock.calls.length - 1;
    let callArgs = mockDayPicker.mock.calls[lastCallIndex][0] as DayPickerProps & {
      mode: 'range';
      selected?: DateRange;
    };
    expect(callArgs.selected).toBeDefined();

    // Clear mock calls to track only new renders
    mockDayPicker.mockClear();

    // Rerender with selected as undefined
    rerender(
      <RangeCalendar
        mode="range-single"
        startMonth={startMonth}
        selected={undefined} // Explicitly pass undefined
        onSelect={() => {}}
      />
    );

    expect(mockDayPicker).toHaveBeenCalled();
    lastCallIndex = mockDayPicker.mock.calls.length - 1;
    callArgs = mockDayPicker.mock.calls[lastCallIndex][0] as DayPickerProps & {
      mode: 'range';
      selected?: DateRange;
    };
    // The dayPickerSelectedProp in RangeCalendar should become undefined
    expect(callArgs.selected).toBeUndefined();

    const mockElementAfterClear = screen.getByTestId('mocked-daypicker');
    expect(mockElementAfterClear).not.toHaveTextContent('Selected From:');
    expect(mockElementAfterClear).not.toHaveTextContent('Selected To:');
    expect(mockElementAfterClear).toMatchSnapshot();
  });

  it('handles basic range selection via click', async () => {
    const user = userEvent.setup();
    const onSelect = vi.fn();

    render(<RangeCalendar {...defaultTestProps} onSelect={onSelect} />);

    // Click on day 5 to start selection
    const day5 = screen.getByTestId('date-5');
    await user.click(day5);

    // Should be called with partial range
    expect(onSelect).toHaveBeenCalledWith({ from: new Date(2025, 5, 5), to: undefined });

    onSelect.mockClear();

    // Click on day 10 to complete selection
    const day10 = screen.getByTestId('date-10');
    await user.click(day10);

    // Should be called with complete range
    expect(onSelect).toHaveBeenCalledWith({
      from: new Date(2025, 5, 5),
      to: new Date(2025, 5, 10),
    });
  });

  it('handles min and max day count constraints', async () => {
    const user = userEvent.setup();
    const onSelect = vi.fn();

    render(<RangeCalendar {...defaultTestProps} min={3} max={5} onSelect={onSelect} />);

    // Click on day 5 to start selection
    const day5 = screen.getByTestId('date-5');
    await user.click(day5);

    expect(onSelect).toHaveBeenCalledWith({ from: new Date(2025, 5, 5), to: undefined });

    onSelect.mockClear();

    // Click on day 6 to create a 2-day range (violates min=3)
    const day6 = screen.getByTestId('date-6');
    await user.click(day6);

    // Should reject the range and make the clicked date the new single selection
    expect(onSelect).toHaveBeenCalledWith({ from: new Date(2025, 5, 6), to: new Date(2025, 5, 6) });
  });

  it('handles disabled dates within a range attempt', async () => {
    const user = userEvent.setup();
    const onSelect = vi.fn();
    const disabledDates = [new Date(2025, 5, 7)]; // Disable June 7th

    render(<RangeCalendar {...defaultTestProps} onSelect={onSelect} disabled={disabledDates} />);

    // Disabled date should not be clickable
    const disabledDay = screen.getByTestId('date-7');
    expect(disabledDay).toBeDisabled();

    // Click on day 5 to start selection
    const day5 = screen.getByTestId('date-5');
    await user.click(day5);

    expect(onSelect).toHaveBeenCalledWith({ from: new Date(2025, 5, 5), to: undefined });

    onSelect.mockClear();

    // Clicking on disabled date should not trigger onSelect
    await user.click(disabledDay);
    expect(onSelect).not.toHaveBeenCalled();
  });

  it('selects range with Enter/Space key', async () => {
    const user = userEvent.setup();
    const onSelect = vi.fn();

    render(<RangeCalendar {...defaultTestProps} onSelect={onSelect} />);

    // Focus and press Enter on day 5
    const day5 = screen.getByTestId('date-5');
    day5.focus();
    await user.keyboard('{Enter}');

    expect(onSelect).toHaveBeenCalledWith({ from: new Date(2025, 5, 5), to: undefined });

    onSelect.mockClear();

    // Focus and press Space on day 10
    const day10 = screen.getByTestId('date-10');
    day10.focus();
    await user.keyboard(' ');

    expect(onSelect).toHaveBeenCalledWith({
      from: new Date(2025, 5, 5),
      to: new Date(2025, 5, 10),
    });
  });

  it('displays min only constraint message', async () => {
    render(
      <RangeCalendar {...defaultTestProps} min={3} selected={{ from: new Date(2025, 5, 10) }} />
    );
    expect(screen.getByTestId('min-constraint-message')).toBeInTheDocument();
    expect(screen.getByTestId('min-constraint-message')).toHaveTextContent(
      'Select at least 3 days'
    );
  });

  it('displays max only constraint message', async () => {
    render(
      <RangeCalendar {...defaultTestProps} max={5} selected={{ from: new Date(2025, 5, 10) }} />
    );
    expect(screen.getByTestId('max-constraint-message')).toBeInTheDocument();
    expect(screen.getByTestId('max-constraint-message')).toHaveTextContent('Select at most 5 days');
  });

  it('displays min and max constraint message', async () => {
    render(
      <RangeCalendar
        {...defaultTestProps}
        min={3}
        max={5}
        selected={{ from: new Date(2025, 5, 10) }}
      />
    );
    expect(screen.getByTestId('min-max-constraint-message')).toBeInTheDocument();
    expect(screen.getByTestId('min-max-constraint-message')).toHaveTextContent(
      'Select between 3 and 5 days'
    );
  });
});
