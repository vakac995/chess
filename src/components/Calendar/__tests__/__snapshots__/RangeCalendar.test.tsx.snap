// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RangeCalendar > clears selection when selected prop is undefined 1`] = `
<div
  aria-label="June 2025"
  data-classnames-root="rdp-root w-full font-sans text-text bg-background transition-colors duration-200 [&_*]:transition-colors [&_*]:duration-200"
  data-start-month="2025-05-31T21:00:00.000Z"
  data-testid="mocked-daypicker"
  role="grid"
>
  <div>
    Mocked DayPicker. Displayed Month:
     
    June 2025
  </div>
  <table
    aria-label="Calendar grid"
  >
    <tbody>
      <tr>
        <td>
          <button
            aria-label="Sun Jun 01 2025"
            data-testid="date-1"
            tabindex="0"
          >
            1
          </button>
        </td>
        <td>
          <button
            aria-label="Mon Jun 02 2025"
            data-testid="date-2"
            tabindex="-1"
          >
            2
          </button>
        </td>
        <td>
          <button
            aria-label="Tue Jun 03 2025"
            data-testid="date-3"
            tabindex="-1"
          >
            3
          </button>
        </td>
        <td>
          <button
            aria-label="Wed Jun 04 2025"
            data-testid="date-4"
            tabindex="-1"
          >
            4
          </button>
        </td>
        <td>
          <button
            aria-label="Thu Jun 05 2025"
            data-testid="date-5"
            tabindex="-1"
          >
            5
          </button>
        </td>
        <td>
          <button
            aria-label="Fri Jun 06 2025"
            data-testid="date-6"
            tabindex="-1"
          >
            6
          </button>
        </td>
        <td>
          <button
            aria-label="Sat Jun 07 2025"
            data-testid="date-7"
            tabindex="-1"
          >
            7
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            aria-label="Sun Jun 08 2025"
            data-testid="date-8"
            tabindex="-1"
          >
            8
          </button>
        </td>
        <td>
          <button
            aria-label="Mon Jun 09 2025"
            data-testid="date-9"
            tabindex="-1"
          >
            9
          </button>
        </td>
        <td>
          <button
            aria-label="Tue Jun 10 2025"
            data-testid="date-10"
            tabindex="-1"
          >
            10
          </button>
        </td>
        <td>
          <button
            aria-label="Wed Jun 11 2025"
            data-testid="date-11"
            tabindex="-1"
          >
            11
          </button>
        </td>
        <td>
          <button
            aria-label="Thu Jun 12 2025"
            data-testid="date-12"
            tabindex="-1"
          >
            12
          </button>
        </td>
        <td>
          <button
            aria-label="Fri Jun 13 2025"
            data-testid="date-13"
            tabindex="-1"
          >
            13
          </button>
        </td>
        <td>
          <button
            aria-label="Sat Jun 14 2025"
            data-testid="date-14"
            tabindex="-1"
          >
            14
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            aria-label="Sun Jun 15 2025"
            data-testid="date-15"
            tabindex="-1"
          >
            15
          </button>
        </td>
        <td>
          <button
            aria-label="Mon Jun 16 2025"
            data-testid="date-16"
            tabindex="-1"
          >
            16
          </button>
        </td>
        <td>
          <button
            aria-label="Tue Jun 17 2025"
            data-testid="date-17"
            tabindex="-1"
          >
            17
          </button>
        </td>
        <td>
          <button
            aria-label="Wed Jun 18 2025"
            data-testid="date-18"
            tabindex="-1"
          >
            18
          </button>
        </td>
        <td>
          <button
            aria-label="Thu Jun 19 2025"
            data-testid="date-19"
            tabindex="-1"
          >
            19
          </button>
        </td>
        <td>
          <button
            aria-label="Fri Jun 20 2025"
            data-testid="date-20"
            tabindex="-1"
          >
            20
          </button>
        </td>
        <td>
          <button
            aria-label="Sat Jun 21 2025"
            data-testid="date-21"
            tabindex="-1"
          >
            21
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            aria-label="Sun Jun 22 2025"
            data-testid="date-22"
            tabindex="-1"
          >
            22
          </button>
        </td>
        <td>
          <button
            aria-label="Mon Jun 23 2025"
            data-testid="date-23"
            tabindex="-1"
          >
            23
          </button>
        </td>
        <td>
          <button
            aria-label="Tue Jun 24 2025"
            data-testid="date-24"
            tabindex="-1"
          >
            24
          </button>
        </td>
        <td>
          <button
            aria-label="Wed Jun 25 2025"
            data-testid="date-25"
            tabindex="-1"
          >
            25
          </button>
        </td>
        <td>
          <button
            aria-label="Thu Jun 26 2025"
            data-testid="date-26"
            tabindex="-1"
          >
            26
          </button>
        </td>
        <td>
          <button
            aria-label="Fri Jun 27 2025"
            data-testid="date-27"
            tabindex="-1"
          >
            27
          </button>
        </td>
        <td>
          <button
            aria-label="Sat Jun 28 2025"
            data-testid="date-28"
            tabindex="-1"
          >
            28
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            aria-label="Sun Jun 29 2025"
            data-testid="date-29"
            tabindex="-1"
          >
            29
          </button>
        </td>
        <td>
          <button
            aria-label="Mon Jun 30 2025"
            data-testid="date-30"
            tabindex="-1"
          >
            30
          </button>
        </td>
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
